src/Analyzers/Core/Analyzers/NamespaceSync/AbstractNamespaceSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+        /// </summary>
+        protected abstract ISyntaxFacts GetSyntaxFacts();
+
+        /// <summary>
+        /// Returns true if the namespace declaration contains one or more partial types with multiple declarations.
+        /// </summary>
+        /// <returns></returns>
+        protected abstract bool ContainsPartialTypeWithMultipleDeclarations(TNamespaceSyntax namespaceDeclaration, SemanticModel semanticModel);
+
+        /// <summary>
+        /// Gets the syntax representing the name of a namespace declaration
+        /// </summary>
+        protected abstract SyntaxNode GetNameSyntax(TNamespaceSyntax namespaceDeclaration);
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            // It's ok to not have a rootnamespace property, but if it's there we want to use it correctly
+            context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace);
+
+            // Project directory is a must to correctly get the relative path and construct a namespace
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || string.IsNullOrEmpty(projectDir))
+            {
+                return;
+            }
+
+            if (context.Node is TNamespaceSyntax namespaceDecl)
+            {
+                var currentNamespace = GetNamespaceName(namespaceDecl, rootNamespace);
+
+                if (IsFileAndNamespaceMismatch(namespaceDecl, rootNamespace, projectDir, out var targetNamespace) &&
+                    IsFixSupported(context.SemanticModel, namespaceDecl, context.CancellationToken))
+                {
+                    var nameSyntax = GetNameSyntax(namespaceDecl);
+                    AbstractNamespaceSyncDiagnosticAnalyzer<TNamespaceSyntax>.ReportDiagnostics(context, Descriptor, nameSyntax.GetLocation(), targetNamespace, currentNamespace);
+                }
+            }
+        }
+
+        private bool IsFixSupported(SemanticModel semanticModel, TNamespaceSyntax namespaceDeclaration, CancellationToken cancellationToken)
+        {
+            var root = namespaceDeclaration.SyntaxTree.GetRoot(cancellationToken);
+
+            // It should not be nested in other namespaces
+            if (namespaceDeclaration
+                .Ancestors()
+                .OfType<TNamespaceSyntax>()
+                .Any())
+            {
+                return false;
+            }
+
+            // It should not contain a namespace
+            var containsNamespace = namespaceDeclaration
+                 .DescendantNodes(n => n is TNamespaceSyntax)
+                 .OfType<TNamespaceSyntax>().Any();
+            if (containsNamespace)
+            {
+                return false;
+            }
```

Comment:
i'm not certain i understnad the logic here.  I think my expectation would be:

1. we grab all the types in the file.  If they're all from the same namespace, compare that namespace to the folder structure.  If they don't match, report hte issue.

Reply:
This is not currently how the sync namespace works in the IDE. IMO, if we change the expectation we should do in both places. @genlu  thoughts?